### PR TITLE
Add Prow plugin for requesting PR reviews

### PR DESCRIPTION
**Pull request by cjwagner**
Similar to the `/assign` and `/unassign` commands,
this command allows the use of `/cc @user` to request a PR review from "user" or `/uncc @user` to remove a PR review request for "user". Multiple users may be listed in a command and different commands may be listed on different lines.

### CODE REVIEW

**Comment by spxtr** [comment]
Can non-org-members already request reviews for their PRs?

`/cc @user` making a review request violates the principle of least astonishment. If our goal is to make these things show up on gubernator then we should make gubernator understand @ mentions instead of hacking around with review requests.

**Comment by fejta** [comment]
I really want to try out this plugin and see how it works.

I base this off two hunches:
a) The `/assign @foo` has a good attach rate
b) Some variation of `cc @bar` is a common pattern within the team.

My belief is that it may be useful to formalize this a bit.

If no one uses the new command and/or it annoys people then we can delete/disable/fix it.

But I really think the best way to get this information -- especially since plugins like this are easy to make -- is to try it out.

File: prow/github/client.go [hunk 1]
```diff
@@ -519,6 +521,81 @@ func (c *Client) UnassignIssue(org, repo string, number int, logins []string) er
...
+	return c.request(&request{
+		method:      http.MethodPost,
+		path:        fmt.Sprintf("%s/repos/%s/%s/pulls/%d/requested_reviewers", c.base, org, repo, number),
+		requestBody: map[string][]string{"reviewers": logins},
+		exitCodes:   []int{201},
+	}, &pr)
+}
+
+func (c *Client) RequestPRReview(org, repo string, number int, logins []string) error {
+	//If any user in the 'logins' slice is not a contributor of the repo, the entire POST will fail
```
Comment (spxtr) [inline prow/github/client.go]: Add a space after `//`, here and elsewhere.

File: prow/github/client.go [hunk 2]
```diff
@@ -519,6 +521,81 @@ func (c *Client) UnassignIssue(org, repo string, number int, logins []string) er
 	return nil
 }
 
+func (c *Client) TryRequestPRReview(org, repo string, number int, logins []string) (int, error) {
```
Comment (spxtr) [inline prow/github/client.go]: Make this private unless you plan on using it elsewhere.

File: prow/github/client.go (hunk 1 above)
Comment (spxtr) [inline prow/github/client.go]: Also, this comment is valuable information that should probably live as a godoc comment above the method name.

File: prow/github/client.go [hunk 3]
```diff
@@ -519,6 +521,81 @@ func (c *Client) UnassignIssue(org, repo string, number int, logins []string) er
...
+	if statusCode == 422 {
+		//failed to set all members of 'logins' as reviewers, try individually.
+		var missing MissingUsers
+		for _, user := range logins {
+			statusCode, err = c.TryRequestPRReview(org, repo, number, []string{user})
+			if statusCode == 422 {
+				//user is not a contributor
+				missing = append(missing, user)
+			} else {
+				if err != nil {
```
Comment (spxtr) [inline prow/github/client.go]: `else if`

File: prow/github/client.go [hunk 4]
```diff
@@ -519,6 +521,81 @@ func (c *Client) UnassignIssue(org, repo string, number int, logins []string) er
...
+	}, &pr)
+}
+
+func (c *Client) RequestPRReview(org, repo string, number int, logins []string) error {
+	//If any user in the 'logins' slice is not a contributor of the repo, the entire POST will fail
+	//without adding any reviewers. The github API response does not specify which user(s) were invalid
+	//so if we fail to request reviews from the members of 'logins' we try to request reviews from
+	//each member individually. We try first with all users in 'logins' for efficiency in the common case.
+	statusCode, err := c.TryRequestPRReview(org, repo, number, logins)
+	if statusCode == 422 {
```
Comment (spxtr) [inline prow/github/client.go]:
```go
if err != nil && statusCode == 422 {
```
Just to be explicit.

File: prow/github/client.go [hunk 5]
```diff
@@ -519,6 +521,81 @@ func (c *Client) UnassignIssue(org, repo string, number int, logins []string) er
...
+		//failed to set all members of 'logins' as reviewers, try individually.
+		var missing MissingUsers
+		for _, user := range logins {
+			statusCode, err = c.TryRequestPRReview(org, repo, number, []string{user})
+			if statusCode == 422 {
+				//user is not a contributor
+				missing = append(missing, user)
+			} else {
+				if err != nil {
+					return err
```
Comment (spxtr) [inline prow/github/client.go]: Consider returning a more useful error in this case.

File: prow/plugins/assign/assign.go [hunk 6]
```diff
@@ -27,23 +27,37 @@ import (
 	"k8s.io/test-infra/prow/plugins"
 )
 
-const pluginName = "assign"
+const (
+	pluginNameAssign = "assign"
+	pluginNameReview = "review"
```
Comment (spxtr) [inline prow/plugins/assign/assign.go]: It should either be in the same plugin as the `/assign` code or in a separate package.

File: prow/plugins/assign/assign.go [hunk 7]
```diff
@@ -27,23 +27,37 @@ import (
...
 	UnassignIssue(owner, repo string, number int, logins []string) error
+
+	RequestPRReview(org, repo string, number int, logins []string) error
+	UnRequestPRReview(org, repo string, number int, logins []string) error
+
+	CreateComment(owner, repo string, number int, comment string) error
 }
 
-type assignEvent struct {
+type Event struct {
```
Comment (spxtr) [inline prow/plugins/assign/assign.go]: Don't export this.

File: prow/github/client.go [hunk 8]
```diff
@@ -519,6 +521,82 @@ func (c *Client) UnassignIssue(org, repo string, number int, logins []string) er
...
+
+// RequestPRReview tries to add the users listed in 'logins' as requested reviewers of the specified PR.
+// If any user in the 'logins' slice is not a contributor of the repo, the entire POST will fail
+// without adding any reviewers. The github API response does not specify which user(s) were invalid
+// so if we fail to request reviews from the members of 'logins' we try to request reviews from
+// each member individually. We try first with all users in 'logins' for efficiency in the common case.
+func (c *Client) RequestPRReview(org, repo string, number int, logins []string) error {
+	statusCode, err := c.tryRequestPRReview(org, repo, number, logins)
+	if err != nil && statusCode == 422 {
+		//failed to set all members of 'logins' as reviewers, try individually.
```
Comment (spxtr) [inline prow/github/client.go]: It's typical in go to put a space after the `//` and start with a capital letter for all comments.

File: prow/github/client.go [hunk 9]
```diff
@@ -519,6 +521,82 @@ func (c *Client) UnassignIssue(org, repo string, number int, logins []string) er
...
+		var missing MissingUsers
+		for _, user := range logins {
+			statusCode, err = c.tryRequestPRReview(org, repo, number, []string{user})
+			if err != nil && statusCode == 422 {
+				//user is not a contributor
+				missing = append(missing, user)
+			} else if err != nil {
+				return fmt.Errorf("Unexpected error or status code while Requesting PR Review. Status: %d, errmsg: %v", statusCode, err)
```
Comment (spxtr) [inline prow/github/client.go]: Error strings should not be capitalized. This one also probably doesn't need to be so verbose.

File: prow/github/client.go [hunk 10]
```diff
@@ -519,6 +521,82 @@ func (c *Client) UnassignIssue(org, repo string, number int, logins []string) er
...
+}
+
+// UnRequestPRReview tries to remove the users listed in 'logins' from the requested reviewers of the
+// specified PR. The github API treats deletions of review requests differently than creations. Specifically, if
+// 'logins' contains a user that isn't a requested reviewer, other users that are valid are still removed.
+// Furthermore, the API response lists the set of requested reviewers after the deletion (unlike request creations),
+// so we can determine if each deletion was successful.
+// The API responds with http status code 200 no matter what the content of 'logins' is.
+func (c *Client) UnRequestPRReview(org, repo string, number int, logins []string) error {
+
```
Comment (spxtr) [inline prow/github/client.go]: Remove this newline.

File: prow/github/client.go [hunk 11]
```diff
@@ -519,6 +521,82 @@ func (c *Client) UnassignIssue(org, repo string, number int, logins []string) er
...
+	return err
+}
+
+// UnRequestPRReview tries to remove the users listed in 'logins' from the requested reviewers of the
+// specified PR. The github API treats deletions of review requests differently than creations. Specifically, if
+// 'logins' contains a user that isn't a requested reviewer, other users that are valid are still removed.
+// Furthermore, the API response lists the set of requested reviewers after the deletion (unlike request creations),
+// so we can determine if each deletion was successful.
+// The API responds with http status code 200 no matter what the content of 'logins' is.
+func (c *Client) UnRequestPRReview(org, repo string, number int, logins []string) error {
```
Comment (spxtr) [inline prow/github/client.go]: I feel like this should be `UnrequestPRReview`, but I'm not sure.

File: prow/plugins/assign/assign.go [hunk 12]
```diff
@@ -104,12 +111,27 @@ func parseLogins(text string) []string {
 	return parts
 }
 
-func handle(gc githubClient, log *logrus.Entry, ae assignEvent) error {
-	if ae.action != "created" && ae.action != "opened" { // Only consider new comments/issues
+func handle(gc githubClient, log *logrus.Entry, e event, isPR bool) error {
```
Comment (spxtr) [inline prow/plugins/assign/assign.go]: Avoid passing a `bool` that is used only for control flow.

File: prow/plugins/assign/assign.go [hunk 13]
```diff
@@ -138,18 +160,64 @@ func handle(gc githubClient, log *logrus.Entry, ae assignEvent) error {
...
+				}
+				return nil
+			}
+			return err
+		}
+	}
+	return nil
+}
+
+func handleReview(gc githubClient, log *logrus.Entry, e event) error {
```
Comment (spxtr) [inline prow/plugins/assign/assign.go]: `handleReview` looks extremely similar to `handleAssign`. Please share code between the two.

File: prow/github/client.go [hunk 14]
```diff
@@ -454,7 +456,7 @@ func (c *Client) RemoveLabel(org, repo string, number int, label string) error {
 type MissingUsers []string
 
 func (m MissingUsers) Error() string {
-	return fmt.Sprintf("cannot assign %s", strings.Join(m, ", "))
+	return fmt.Sprintf("cannot assign or request review from %s", strings.Join(m, ", "))
```
Comment (fejta) [inline prow/github/client.go]: Are we attempting to assign or are we attempting to request a review?

This change makes the error more confusing. Can we fix this to say what isn't happening corrrectly?

File: prow/github/client.go [hunk 15]
```diff
@@ -519,6 +521,82 @@ func (c *Client) UnassignIssue(org, repo string, number int, logins []string) er
...
+	var pr PullRequest
+	return c.request(&request{
+		method:      http.MethodPost,
+		path:        fmt.Sprintf("%s/repos/%s/%s/pulls/%d/requested_reviewers", c.base, org, repo, number),
+		requestBody: map[string][]string{"reviewers": logins},
+		exitCodes:   []int{201},
+	}, &pr)
+}
+
+// RequestPRReview tries to add the users listed in 'logins' as requested reviewers of the specified PR.
```
Comment (fejta) [inline prow/github/client.go]: Really? /assign works differently, which is surprising.

File: prow/github/client.go [hunk 16]
```diff
@@ -519,6 +521,82 @@ func (c *Client) UnassignIssue(org, repo string, number int, logins []string) er
...
+// If any user in the 'logins' slice is not a contributor of the repo, the entire POST will fail
+// without adding any reviewers. The github API response does not specify which user(s) were invalid
+// so if we fail to request reviews from the members of 'logins' we try to request reviews from
+// each member individually. We try first with all users in 'logins' for efficiency in the common case.
+func (c *Client) RequestPRReview(org, repo string, number int, logins []string) error {
```
Comment (fejta) [inline prow/github/client.go]: PR seems redundant.

File: prow/github/client.go [hunk 17]
```diff
@@ -519,6 +521,81 @@ func (c *Client) UnassignIssue(org, repo string, number int, logins []string) er
...
+	statusCode, err := c.tryRequestPRReview(org, repo, number, logins)
+	if err != nil && statusCode == 422 {
```
Comment (fejta) [inline prow/github/client.go]: Do you know about the http constants for 422?

File: prow/plugins/assign/assign_test.go [hunk 18]
```diff
@@ -202,30 +247,127 @@ func TestAssignComment(t *testing.T) {
...
+			body:      "/cc @cjwagner @merlin",
+			commenter: "rando",
+			requested: []string{"cjwagner", "merlin"},
+		},
+		{
+			name:        "unrequest buddies",
+			action:      "created",
+			body:        "/uncc @ashitaka @eboshi",
+			commenter:   "san",
+			unrequested: []string{"ashitaka", "eboshi"},
```
Comment (fejta) [inline prow/plugins/assign/assign_test.go]: :rofl:

**Comment by fejta** [comment]
LGTM, deferring to Joe for his more rigorous review

File: prow/github/client.go (hunk 15 above)
Comment (spxtr) [inline prow/github/client.go]: PR Review Requests are still in the "developer preview" window, and I'm guessing this is the sort of behavior that will be broken at some point.

File: prow/github/client.go (hunk 16 above)
Comment (spxtr) [inline prow/github/client.go]: I'm ambivalent about this one.

File: prow/github/client.go (hunk 17 above)
Comment (spxtr) [inline prow/github/client.go]: `http.StatusUnprocessableEntity`

File: prow/github/client.go [hunk 19]
```diff
@@ -519,6 +521,81 @@ func (c *Client) UnassignIssue(org, repo string, number int, logins []string) er
...
+		//Failed to set all members of 'logins' as reviewers, try individually.
```
Comment (spxtr) [inline prow/github/client.go]: Space after `//` :)

File: prow/plugins/assign/assign.go [hunk 20]
```diff
@@ -104,58 +127,155 @@ func parseLogins(text string) []string {
...
+	if err := gc.CreateComment(e.org, e.repo, e.number, plugins.FormatResponseRaw(e.body, e.url, e.login, msg)); err != nil {
+		return fmt.Errorf("comment err: %v", err)
 	}
+	return nil
+}
 
-	matches := assignRe.FindAllStringSubmatch(ae.body, -1)
+func combineErrors(err1, err2 error) error {
+	if err1 != nil && err2 != nil {
+		return fmt.Errorf("Two errors: 1)%v 2)%v.", err1, err2)
```
Comment (spxtr) [inline prow/plugins/assign/assign.go]: error strings start with a lowercase letter.

File: prow/plugins/assign/assign.go [hunk 21]
```diff
@@ -104,58 +127,155 @@ func parseLogins(text string) []string {
...
-				return nil
+				return h.handleAddFailure(mu)
 			}
 			return err
 		}
 	}
 	return nil
 }
+
+type handler interface {
```
Comment (spxtr) [inline prow/plugins/assign/assign.go]: I think you can do away with the interface and go with a single struct that contains those functions. Note that the GitHub client function signatures for the two implementations of the functions are identical, so RequestReview looks just like AssignIssue if you squint at it.

File: prow/github/client.go [hunk 22]
```diff
@@ -473,22 +478,25 @@ func (c *Client) AssignIssue(org, repo string, number int, logins []string) erro
...
+	if len(missing.Users) > 0 {
 		return missing
 	}
 	return nil
 }
 
-type ExtraUsers []string
+type ExtraUsers struct {
+	Users     []string
+	errPrefix string
```
Comment (fejta) [inline prow/github/client.go]: I'd favor prefix

File: prow/plugins/assign/assign.go (hunk 21 above)
Comment (cjwagner) [inline prow/plugins/assign/assign.go]: I switched the interface to a struct and made the other changes you requested as well.

**Comment by fejta** [comment]
LGTM

**Comment by spxtr** [comment]
LGTM, please run `./bump.sh hook` and commit the results to the PR. Then we can merge and deploy.

**Pull request merged**